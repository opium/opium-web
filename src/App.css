.ThumbnailList > .Thumbnail {
    display: inline-block;
    width: 200px;
}
