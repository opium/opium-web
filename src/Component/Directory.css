.ThumbnailList {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-items: stretch;
}

.Footer {
    text-align: right;
}

.DirectoryHeaderWithBanner {
    height: 400px;
    background-position: center center;
    background-size: cover;
    margin-bottom: 80px;
}

.DirectoryHeaderTitle {
    margin-top: 70px;
    color: #fff;
    text-align: center;
    font-size: 54px;
}

.Back {
    display: inline-block;
    padding: 1em;
    font-size: 1.2em;
    color: #899eaf;
    text-decoration: none;
}

.Back:hover {
    text-decoration: underline;
}
