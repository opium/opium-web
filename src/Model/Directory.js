import { List, Map, Record } from 'immutable';
import { entityFactory, mapEntityRelationShips } from './Factory';

class Directory extends Record({
  id: null,
  name: null,
  slug: null,
  children: [],
  _links: {},
  _embedded: {},
  imageLines: [],
  parent: null,
  directoryThumbnail: null,
}) {
  constructor(val) {
    const data = val;
    if (val.image_lines) {
      data.imageLines = List(val.image_lines.map(line =>
        Map(Object.keys(line).map(key => [key, line[key]]))
      ));
    }

     data.directoryThumbnail = val._embedded &&
       val._embedded.directory_thumbnail &&
       entityFactory(val._embedded.directory_thumbnail);

    return mapEntityRelationShips(super(data), data);
  }

  getChildById(id) {
    return this.children.find(item => item.id === parseInt(id, 10));
  }

  getChildrenSize() {
    return this.children.size || 0;
  }

  getPositionBounds() {
    let bounds = Map();
    const children = this.children.filter(c => !!c.position);

    if (children.size === 1) {
      const child = children.first().position;

      return Map({
        top: child.get('lat') + 1,
        bottom: child.get('lat') - 1,
        left: child.get('lng') - 1,
        right: child.get('lng') + 1,
      });
    }

    for (const child of children) {
      const latitude = child.position.get('lat');
      const longitude = child.position.get('lng');

      if (!bounds.get('top') || bounds.get('top') < latitude) {
        bounds = bounds.set('top', latitude);
      }
      if (!bounds.get('bottom') || bounds.get('bottom') > latitude) {
        bounds = bounds.set('bottom', latitude);
      }
      if (!bounds.get('right') || bounds.get('right') < longitude) {
        bounds = bounds.set('right', longitude);
      }
      if (!bounds.get('left') || bounds.get('left') > longitude) {
        bounds = bounds.set('left', longitude);
      }
    }

    if (!bounds.get('top')) {
      return null;
    }

    const latDelta = (bounds.get('top') - bounds.get('bottom')) / 4;
    const lngDelta = (bounds.get('right') - bounds.get('left')) / 4;

    return bounds.merge({
      top: bounds.get('top') + latDelta,
      bottom: bounds.get('bottom') - latDelta,
      left: bounds.get('left') - lngDelta,
      right: bounds.get('right') + lngDelta,
    });
  }

  // hack until https://github.com/mapado/rest-client-js-sdk/issues/11 is resolved
  get(key) {
    if (key === '@id') {
      return `/v1/directories/${this.id}`;
    }

    return super.get(key);
  }
}

export default Directory;
