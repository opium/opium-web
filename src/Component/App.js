/* global container */
import React, { Component, PropTypes } from 'react';
import './App.css';
import Thumbnail from './Thumbnail';

class App extends Component {
  static propTypes = {
    directory: PropTypes.object,
  }

  constructor(props) {
    super(props);

    this.renderOneLine = this.renderOneLine.bind(this);
  }

  renderOneLine(line) {
    const directory = this.props.directory;

    return line.entrySeq().map(([id, thumbnail]) =>
      <Thumbnail
        key={id}
        title={directory.getChildById(id).name}
        image={thumbnail.thumbs}
      />
    );
  }

  render() {
    const directory = this.props.directory;

    return (
      <div>
        <div className="ThumbnailList">
          {directory && directory.imageLines.map(this.renderOneLine)}
        </div>

        <footer className="Footer">
          Total number of items: {directory && directory.getChildrenSize()}
        </footer>
      </div>
    );
  }
}

export default App;
