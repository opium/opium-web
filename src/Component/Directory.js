import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import Helmet from 'react-helmet';
import cn from 'classnames';
import ChevronLeft from 'react-icons/lib/ti/chevron-left';
import './Directory.css';
import Thumbnail from './Thumbnail';
import File from '../Model/File';
import ImageWithLoader from './ImageWithLoader';

class DirectoryHeader extends Component {
  static propTypes = {
    directory: PropTypes.object.isRequired,
    viewportWidth: PropTypes.number.isRequired,
  };

  constructor(props) {
    super(props);

    this.state = {
      backgroundLoaded: null,
    };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.directory !== this.props.directory) {
      this.setState({
        backgroundLoaded: null,
      });
    }
  }

  render() {
    const { directory, viewportWidth } = this.props;

    if (!directory.parent && !directory.directoryThumbnail) {
      return null;
    }


    const styles = {};
    if (directory.directoryThumbnail) {
      if (this.state.backgroundLoaded) {
        const backgroundImage = directory.directoryThumbnail.generateCrop(viewportWidth, 400);
        styles.backgroundImage = `linear-gradient(rgba(0, 0, 0, 0) 50%,
          rgba(0, 0, 0, 0.8)),
          url(${backgroundImage})`;
      }
    }

    return (
      <header
        className={cn(
          directory.directoryThumbnail && 'DirectoryHeader--WithBanner',

        )}
      >
        {directory.directoryThumbnail &&
          <div>
            <ImageWithLoader
              src={directory.directoryThumbnail.generateCrop(viewportWidth, 400)}
              onLoad={() => this.setState({ backgroundLoaded: true })}
              loaderProps={{ className: 'DirectoryHeader__Image' }}
            />
            <div
              className={cn(
                'DirectoryHeader__Image',
                this.state.backgroundLoaded && 'DirectoryHeader__Image--Opaque'
              )}
              style={styles}
            />
          </div>
        }

        {directory.parent &&
          <Link to={`/${directory.parent.slug}`} className="DirectoryHeader__Back">
            <ChevronLeft />
            Back to albums
          </Link>
        }

        {directory.directoryThumbnail &&
          <h1 className="DirectoryHeader__Title">
            {directory.name}
          </h1>
        }
      </header>
    );
  }
};

class Directory extends Component {
  static propTypes = {
    directory: PropTypes.object,
    findDirectory: PropTypes.func.isRequired,
    slug: PropTypes.string.isRequired,
    viewportWidth: PropTypes.number.isRequired,
  }

  constructor(props) {
    super(props);

    this.renderOneLine = this.renderOneLine.bind(this);
  }

  componentDidMount() {
    this.props.findDirectory(this.props.slug);
  }

  componentDidUpdate(prevProps) {
    if (this.props.slug !== prevProps.slug) {
      this.props.findDirectory(this.props.slug);
    }
  }

  renderOneLine(line) {
    const directory = this.props.directory;

    return line.entrySeq().map(([id, thumbnail]) => {
      const child = directory.getChildById(id);
      const childSlug = child instanceof File ?
        `/${directory.slug}/${child.slug}` :
        `/${child.slug}`;

      return (
        <Link to={childSlug}>
          <Thumbnail
            key={id}
            title={child.name}
            image={thumbnail.thumbs}
          />
        </Link>
      );
    });
  }

  render() {
    if (this.props.children) {
      return this.props.children;
    }

    const directory = this.props.directory;

    if (!directory) {
      return <div />;
    }

    return (
      <div>
        <Helmet title={directory.name} />

        <DirectoryHeader
          directory={directory}
          viewportWidth={this.props.viewportWidth}
        />

        <div className="Directory__ThumbnailList">
          {directory.imageLines.map(this.renderOneLine)}
        </div>

        <footer className="DirectoryFooter">
          Total number of items: {directory.getChildrenSize()}
        </footer>
      </div>
    );
  }
}

export default Directory;
