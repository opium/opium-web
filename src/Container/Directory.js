import { connect } from 'react-redux';
import { Set } from 'immutable';
import Directory from '../Component/Directory';
import { find } from '../Action/DirectoryAction';
import { loadImage } from '../Action/ImageAction';

function mapStateToProps(state, ownProps) {
  const viewportWidth = document.body.clientWidth;
  const loadedImages = state.imageLoader.get('loadedImages') || Set();
  const directory = state.opium.get('currentDirectory');
  const hasBackground = !!(directory && directory.directoryThumbnail);
  const directoryImage = hasBackground && directory.directoryThumbnail.generateCrop(viewportWidth, 400);

  return {
    directory,
    slug: ownProps.params.directorySlug || '',
    hasBackground,
    backgroundLoaded: directory && loadedImages.has(directoryImage),
    backgroundImage: directoryImage,
  }
};

const mapDispatchToProps = {
  findDirectory: find,
  loadImage,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Directory);
