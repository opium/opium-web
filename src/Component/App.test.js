import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import Directory from '../Model/Directory';

it('renders without crashing without directory', () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders an empty directory', () => {
  const directory = new Directory({
    name: 'foo',
  });
  const component = renderer.create(<App directory={directory} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});

it('renders a non empty directory', () => {
  const directory = new Directory({
    name: 'foo',
    children: [
      {
        id: 1,
        name: "Alps",
        slug: "alps",
      },
      {
        id: 2,
        name: "Pacific Ocean",
        slug: "pacific-ocean",
      }
    ],
    image_lines: [
      {
        "1": {
          thumbs: 'http://demo.opium.sitioweb.fr/2909-vallon-moy-res-jpg/thumbs/200-200',
        },
        "2": {
          thumbs: 'http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg/thumbs/200-200',
        }
      }
    ],
  });
  const component = renderer.create(<App directory={directory} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});
