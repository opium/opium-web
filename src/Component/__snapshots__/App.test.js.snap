exports[`test renders a non empty directory 1`] = `
<div>
  <div
    className="ThumbnailList">
    <article
      className="Thumbnail"
      style={
        Object {
          "backgroundImage": "url(http://demo.opium.sitioweb.fr/2909-vallon-moy-res-jpg/thumbs/200-200)"
        }
      }>
      <span
        className="Caption">
        Alps
      </span>
    </article>
    <article
      className="Thumbnail"
      style={
        Object {
          "backgroundImage": "url(http://demo.opium.sitioweb.fr/2010-mavericks-competition-edit1-jpg/thumbs/200-200)"
        }
      }>
      <span
        className="Caption">
        Pacific Ocean
      </span>
    </article>
  </div>
  <footer
    className="Footer">
    Total number of items: 
    2
  </footer>
</div>
`;

exports[`test renders an empty directory 1`] = `
<div>
  <div
    className="ThumbnailList" />
  <footer
    className="Footer">
    Total number of items: 
    0
  </footer>
</div>
`;

exports[`test renders without crashing without directory 1`] = `
<div>
  <div
    className="ThumbnailList" />
  <footer
    className="Footer">
    Total number of items: 
  </footer>
</div>
`;
